drt_1d.py

```diff
@@ -32,47 +34,152 @@ def construct_source_term(num_voxels:int, source_voxel_index:int, source_strengt
+    if ri < 0 or ri > len(sigma_s) * voxel_size or rj < 0 or rj > len(sigma_s) * voxel_size:
+        raise ValueError("Positions must be within the spatial domain defined by the voxels.")
     
-    min_index = min(ri, rj)
-    max_index = max(ri, rj)
-    distance = abs(ri - rj) * voxel_size
-    tau = np.sum(sigma_s[min_index:max_index]) * voxel_size
-    streaming_operator = np.exp(-tau) / (4 * np.pi * distance**2)
+    distance = abs(ri - rj)
+    if distance == 0:
+        raise ValueError(f"The positions of (ri,start_index):({ri},{start_index}) and (rj,end_index):({rj},{end_index}) are the same value and will result in an infinite streaming operator.")
+    tau = 0
+    
+    if start_index == end_index:
+        # Same voxel calculation
+        tau = sigma_s[start_index] * distance
+        logging.debug("Intra-voxel tau: %f", tau)
+    else:
+        # Multiple voxel calculation
+        if ri < rj:
+            # Direction is to the right (start_index and ri are on the left)
+            
+            # Calculate left fraction
+            if ri == start_index*voxel_size:
+                # ri is on the left-most point of the voxel
+                left_distance = voxel_size
+            elif start_index*voxel_size < ri and ri < (start_index+1)*voxel_size:
+                # ri is between the voxel edges
+                left_distance = (start_index+1)*voxel_size - ri
+            elif ri == (start_index+1)*voxel_size:
+                # ri is on the right-most point of the voxel
+                left_distance = 0
+            else:
+                # ri is not within or on the edges of this voxel
+                raise ValueError(f"Invalid position (ri: {ri}) and voxel index (start_index: {start_index}) combination.")
+            left_tau = sigma_s[start_index] * left_distance
+            
+            # Calculate right fraction
+            if rj == end_index*voxel_size:
+                # rj is on the left-most point of the voxel
+                right_distance = 0
+            elif end_index*voxel_size < rj and rj < (end_index+1)*voxel_size:
+                # rj is beteween the voxel edges
+                right_distance = rj - (end_index*voxel_size)
+            elif rj == (end_index+1)*voxel_size:
+                # rj is on the right-most point of the voxel
+                right_distance = voxel_size
+            else:
+                # rj is not within or on the edges of this voxel
+                raise ValueError(f"Invalid position (rj: {rj}) and voxel index (end_index: {end_index}) combination.")
+            right_tau = sigma_s[end_index] * right_distance
+            
+            # Sum full voxels
+            center_tau = sum(sigma_s[start_index + 1:end_index]) * voxel_size
+            
+            logging.debug("Direction is to the right. Left_tau: %f, Right_tau: %f, Center_tau: %f", left_tau, right_tau, center_tau)
+            tau = left_tau + right_tau + center_tau
```

Comment:
Slightly simpler to do:
```suggestion
            path = [left_distance] + [voxel_size] * (end_index - start_index - 1) + [right_distance]
            # Sum full voxels
            tau = np.dot(sigma_s[start_index:end_index+1],  np.array(path))
```